---
title: "02 QC"
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: "../../transcriptomics-benchmarks.bib"
biblio-style: apalike
link-citations: true
output:
  html_document:
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding,
      output_file = "../../results/habib-2017-snRNAseq/02-qc.html")})
---

```{r, include = FALSE}
#    This file is part of transcriptomics-benchmarks.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(fig.width = 8.5, fig.height = 7)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of the [Transcriptomics Benchmarks Project](https://github.com/eturkes/transcriptomics-benchmarks).*

```{r}
data_str <- "habib-2017-snRNAseq"
data_str_sub <- unlist(strsplit(data_str, "-"))[1]
assets_dir <- file.path(getwd(), "..", "..", "assets", data_str)
results_dir <- file.path(getwd(), "..", "..", "results", data_str)

packages <- c(
  "conflicted", "SingleCellExperiment", "magrittr", "dplyr", "ggplot2", "ggrepel", "S4Vectors",
  "SummarizedExperiment", "DropletUtils", "scran", "BiocSingular", "scater", "Rtsne", "svd",
  "SC3", "DT", "data.table", "Seurat", "uwot", "viridis", "biomaRt", "robustbase", "edgeR")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))

conflict_prefer("which", "BiocGenerics")
options(stringsAsFactors = FALSE)

# Create a unique cache and results data directory for each iterated section.
if (!dir.exists(file.path(assets_dir, "cache", "02"))) {
  dir.create(file.path(assets_dir, "cache", "02"), recursive = TRUE)}
if (!dir.exists(file.path(results_dir, "data", "02"))) {
  dir.create(file.path(results_dir, "data", "02"), recursive = TRUE)}

# ggplot2 function providing custom aesthetics and automatic placement of categorical labels.
# For continuous data, a colorbar is implemented.
dim_red_plot <- function(data, x, y, col, type) {
  gg <- ggplot(data, aes_string(x = x, y = y, color = col)) +
    geom_point(alpha = 0.35, stroke = 0.05, shape = 21, aes_string(fill = col)) +
    theme_classic() +
    theme(
      legend.position = "right", plot.title = element_text(hjust = 0.5),
      legend.title = element_blank()) +
    guides(color = guide_legend(override.aes = list(alpha = 1)))
    if (type == "cat") {
      gg <- gg + geom_label_repel(data = label_df2, aes(label = label), show.legend = FALSE)
    } else if (type == "cont") {
      gg <- ggplot(data, aes_string(x = x, y = y)) +
        geom_point(alpha = 0.35, stroke = 0.05, aes_string(color = col)) +
        theme_classic() +
        theme(
          legend.position = "right", plot.title = element_text(hjust = 0.5),
          legend.title = element_blank()) +
        scale_colour_viridis()}
  gg}

# Adds download buttons and horizontal scrolling.
datatable_custom <- function(dt) {
  datatable(
    dt, extensions = "Buttons",
    options = list(
      scrollX = TRUE, dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(extend = "collection", buttons = c("csv", "excel", "pdf"), text = "Download"))))}
```

We use the SCE object from the analysis in [github.com/eturkes/habib-2017-snRNAseq](https://github.com/eturkes/habib-2017-snRNAseq).

```{r}
sce <- readRDS(file.path(assets_dir, "sce_orig.rds"))
sce
```

# QC Metrics

We add standard identifiers to our gene symbols for accurate identification of mitochondrial genes and calculate QC metrics.

```{r}
attributes <- c(
  "external_gene_name", "ensembl_gene_id", "entrezgene_id",
  "hgnc_symbol", "description", "chromosome_name")
```

```{r, cache = TRUE}
rds <- file.path(assets_dir, "cache", "02", "gene_anno.rds")
if (file.exists(rds)) {
  gene_anno <- readRDS(rds)
} else {
  mart <- useEnsembl(biomart = "ensembl", GRCh = 37, dataset = "hsapiens_gene_ensembl")
  gene_anno <- getBM(
    attributes = attributes, filters = "external_gene_name", values = rownames(sce), mart = mart)
  saveRDS(gene_anno, rds)
  rm(mart)}
```

Before merging them into the SCE object, the annotation must be cleaned by removing genes not in the RNAseq dataset, irrelevant annotations, duplicate genes, and genes that are not annotated.

```{r}
# Remove genes not in the RNAseq dataset.
remove_genes <- which(!gene_anno$external_gene_name %in% rownames(sce))
gene_anno <- gene_anno[-remove_genes, ]

# Remove annotations to scaffolds, assembly patches, and alternative loci.
chromosomes <- c(1:22, "X", "Y", "MT")
gene_anno <- gene_anno[which(gene_anno$chromosome_name %in% chromosomes), ]

# Remove duplicates.
dup <- table(gene_anno$external_gene_name)
dup <- sort(dup[dup > 1], decreasing = TRUE)
dup <- which(gene_anno$external_gene_name %in% names(dup))
gene_anno2 <- gene_anno[dup, ]
gene_anno2 <- gene_anno2[which(gene_anno2$hgnc_symbol == gene_anno2$external_gene_name), ]
gene_anno2 <- distinct(gene_anno2, external_gene_name, .keep_all = TRUE)
gene_anno <- rbind(gene_anno[-dup, ], gene_anno2)

# Remove missing.
keep_genes <- match(gene_anno$external_gene_name, rownames(sce))
sce <- sce[keep_genes, ]
rowData(sce) <- gene_anno
rm(gene_anno, gene_anno2)
names(rowData(sce))
```

```{r}
is_mito <- which(rowData(sce)$chromosome_name == "MT")
sce <- addPerCellQC(sce, subsets = list(Mito = is_mito), BPPARAM = MulticoreParam())
names(colData(sce))
```

# With Fixed Thresholds

```{r}
lib_size <- colData(sce)$sum < 1e5
num_exprs <- colData(sce)$detected < 5e3
mito <- colData(sce)$subsets_Mito_percent > 10
discard <- lib_size | num_exprs | mito

datatable_custom(data.table(
    low_lib_size = sum(lib_size), low_n_features = sum(num_exprs),
    high_subsets_Mito_percent = sum(mito), discard = sum(discard)))
```

We see that these thresholds are too strict and would remove everything.

# Adaptive Thresholds

```{r}
remove <- quickPerCellQC(sce, percent_subsets = "subsets_Mito_percent")
datatable_custom(t(colSums(as.matrix(remove))))
```

Conversely, using this approach we get larger, and only, removals by mitochondrial proportion.

# Adaptive within Batch

```{r}
remove <- quickPerCellQC(sce, percent_subsets = "subsets_Mito_percent", batch = sce$cell_id_stem)
sce$discard <- remove$discard
datatable_custom(t(colSums(as.matrix(remove))))
```

As expected, the number removed increases as the MAD is no longer inflated by differences in sequencing depth between batches.
We plot some additional metrics regarding this method.

```{r, fig.width = 11}
gridExtra::grid.arrange(
  plotColData(sce, x = "cell_id_stem", y = "sum", colour_by = "discard", theme_size = 6) +
    scale_y_log10() + ggtitle("Total count"),
  plotColData(sce, x = "cell_id_stem", y = "detected", colour_by = "discard", theme_size = 6) +
    scale_y_log10() + ggtitle("Detected features"),
  plotColData(
    sce, x = "cell_id_stem", y = "subsets_Mito_percent", colour_by="discard", theme_size = 6) +
    ggtitle("Mito percent"),
  ncol = 1)
```

```{r}
lost <- calculateAverage((sce)[ , !sce$discard])
kept <- calculateAverage((sce)[ , sce$discard])
logged <- edgeR::cpm(cbind(lost, kept), log = TRUE, prior.count = 2)
logFC <- logged[ , 1] - logged[ , 2]
abundance <- rowMeans(logged)
rm(logged)

plot(abundance, logFC, xlab = "Average Count", ylab="LogFC (lost/kept)", pch = 16)
points(abundance[is_mito], logFC[is_mito], col = "dodgerblue", pch = 16)
```

We see by the last plot that the removals do not correlate with the upregulation of certain genes, which provides confidence that we are not removing a biologically interesting cell-type population.

# robustbase

As an alternate approach, `robustbase` quantifies outlyingness using their own QC methods.

```{r}
stats <- cbind(
  log10(colData(sce)$sum), log10(colData(sce)$detected), colData(sce)$subsets_Mito_percent)
outlying <- adjOutlyingness(stats, only.outlyingness = TRUE)
multi_outlier <- isOutlier(outlying, type = "higher")
rm(stats)
summary(multi_outlier)
```

# Empty Droplets

In a standard droplet experiment, empty droplets with few transcripts or ambient RNA are removed, but it appears this has been performed already.

```{r}
bcrank <- barcodeRanks(counts(sce))

# Only show unique points for plotting speed.
uniq <- !duplicated(bcrank$rank)

plot(
  bcrank$rank[uniq], bcrank$total[uniq], log = "xy",
  xlab = "Rank", ylab = "Total UMI count", cex.lab = 1.2)
abline(h = metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
abline(h = metadata(bcrank)$knee, col = "dodgerblue", lty = 2)
legend(
  "bottomleft", legend = c("Inflection", "Knee"), col = c("darkgreen", "dodgerblue"),
  lty = 2, cex = 1.2)
```

# Seurat

```{r}
seurat <- as.Seurat(sce, data = NULL)
VlnPlot(seurat, features = c("detected", "sum", "subsets_Mito_percent"), ncol = 3)
plot1 <- FeatureScatter(seurat, feature1 = "sum", feature2 = "subsets_Mito_percent")
plot2 <- FeatureScatter(seurat, feature1 = "sum", feature2 = "detected")
rm(bcrank)
CombinePlots(plots = list(plot1, plot2))
```

# Removal

We subset the SCE object using the adaptive within batch approach and the Seurat object using the standard Seurat approach to be saved for downstream analysis.

```{r}
sce <- sce[ , !sce$discard]
seurat <- subset(seurat, subset = detected > 200 & detected < 2500 & subsets_Mito_percent < 5)
rm(plot1, plot2)
dim(sce)
dim(seurat)
```

# References

This is the concluding section of the document. Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r}
saveRDS(sce, file.path(results_dir, "data", "02", "sce.rds"))
saveRDS(sce, file.path(results_dir, "data", "02", "seurat.rds"))

sessionInfo()
```
